airflow/jobs/scheduler_job_runner.py

```diff
@@ -1659,7 +1659,17 @@ def _schedule_dag_run(
                     .where(TI.state.in_(State.unfinished))
                 )
                 for task_instance in unfinished_task_instances:
-                    task_instance.state = TaskInstanceState.SKIPPED
+                    # If the DAG is set to kill the task instances on timeout, we call on_kill
+                    # otherwise we just skip the task instance
+                    if dag.call_on_kill_on_dagrun_timeout:
+                        try:
+                            task = dag.get_task(task_instance.task_id)
+                            task.on_kill()
+                            task_instance.state = TaskInstanceState.FAILED
+                        except Exception as e:
```

Comment:
Should this have a specific exception handling for `AttributeError` in the case where the operator doesn't have an on_kill method defined? In this case, I think we'd want to `self.log.info("on_kill method does not exist for task %s: %s", task_instance, e).

What do you think?

Reply:
Or maybe something like...

```python
if hasattr(task, 'on_kill'):
    try:
        ...
else:
    self.log.warn("Task %s was configured to execute its on_kill method after a DAG run timeout, but it does not have the method 'on_kill' method defined", task_instance)
```

Reply:
If there can be a case of an operator without an `on_kill`, then definitely we should check for that. I think using `hasattr` to check for it before getting the task and using `on_kill` is better to prevent the AttributeError. I will make this change.